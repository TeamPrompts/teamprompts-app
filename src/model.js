import { SEPARATOR } from './constants';

const model = {
  examples: ['San Jose'],
  prompts: ['[place]'],
  source: `Do you know the way to ${SEPARATOR}?`
};

export default model;
