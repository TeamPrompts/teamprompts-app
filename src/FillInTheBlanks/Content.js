import classnames from 'classnames';
import React, { useReducer } from 'react';
import InputText from '../components/InputText/InputText';
import { BLANK, SEPARATOR, modes } from '../constants';
import makeInitialState from './makeInitialState';
import reducer from './reducer';

function getValues(mode, { examples, prompts, source }) {
  let values = [];
  const blanks = source.match(/__BLANK__/g).map(() => BLANK); // INFO: not sure if I can interpolate inside a regex
  switch (mode) {
    case modes.blanks:
    default:
      values = blanks;
      break;
    case modes.examples:
      values = examples;
      break;
    case modes.input:
      if (prompts) {
        values = prompts;
      } else {
        values = blanks;
      }
      break;
    case modes.prompts:
      values = prompts;
      break;
  }
  return values;
}

function interpolate({ dispatch, mode, source, state, values }) {
  const text = source.slice();
  const parts = text.split(SEPARATOR);
  const result = [];
  parts.forEach((part, index) => {
    result.push(part);
    if (values[index]) {
      let element;
      if (mode === modes.input) {
        const hint = values[index];
        element = (
          <InputText
            dispatch={dispatch}
            hint={hint}
            id={index}
            key={index}
            value={state[index]}
          />
        );
      } else {
        element = (
          <span
            className={classnames({
              'bg-yellow-200': mode === modes.blanks,
              'bg-green-200': mode === modes.examples,
              'bg-blue-200': mode === modes.prompts
            })}
            key={index}
          >
            {values[index]}
          </span>
        );
      }
      result.push(element);
    }
  });
  return result;
}

function Content({ fitb, mode }) {
  const [state, dispatch] = useReducer(reducer, makeInitialState(fitb.source));
  const values = getValues(mode, fitb);
  return interpolate({ dispatch, mode, source: fitb.source, state, values });
}

export default Content;
