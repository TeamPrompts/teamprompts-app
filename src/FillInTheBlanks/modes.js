const modes = {
  blanks: 'blanks',
  examples: 'examples',
  input: 'input',
  prompts: 'prompts'
};

export default modes;
