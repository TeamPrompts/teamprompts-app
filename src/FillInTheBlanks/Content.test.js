import React from 'react';
import { create } from 'react-test-renderer';
import model from '../model';
import Content from './Content';
import modes from './modes';

describe('Content', () => {
  const scenarios = [
    {
      description: modes.blanks,
      props: {
        mode: modes.blanks,
        model
      }
    },
    {
      description: modes.examples,
      props: {
        mode: modes.examples,
        model
      }
    },
    {
      description: `${modes.input} w/ prompts`,
      props: {
        mode: modes.input,
        model
      }
    },
    {
      description: `${modes.input} w/o prompts`,
      props: {
        mode: modes.input,
        model: (() => {
          const { prompts, ...rest } = model;
          return rest;
        })()
      }
    },
    {
      description: modes.prompts,
      props: {
        mode: modes.prompts,
        model
      }
    }
  ];

  scenarios.forEach(({ description, props }) => {
    it(description, () => {
      const tree = create(<Content {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
