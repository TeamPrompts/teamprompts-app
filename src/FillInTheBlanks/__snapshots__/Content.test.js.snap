// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content blanks 1`] = `
Array [
  "Do you know the way to ",
  <span
    className="bg-yellow-200"
  >
    ________________
  </span>,
  "?",
]
`;

exports[`Content examples 1`] = `
Array [
  "Do you know the way to ",
  <span
    className="bg-green-200"
  >
    San Jose
  </span>,
  "?",
]
`;

exports[`Content input w/ prompts 1`] = `
Array [
  "Do you know the way to ",
  <input
    placeholder="[place]"
    type="text"
  />,
  "?",
]
`;

exports[`Content input w/o prompts 1`] = `
Array [
  "Do you know the way to ",
  <input
    placeholder="________________"
    type="text"
  />,
  "?",
]
`;

exports[`Content prompts 1`] = `
Array [
  "Do you know the way to ",
  <span
    className="bg-blue-200"
  >
    [place]
  </span>,
  "?",
]
`;
